File: languages/go/projects/04-expert/solutions/main.go
package main

import (
	"fmt"
	"time"
)

type Step struct {
	name           string
	processedCount int
}

func NewStep(name string) *Step {
	return &Step{name: name, processedCount: 0}
}

func (s *Step) Run() {
	s.processedCount++
}

func main() {
	const jobCount = 3

	steps := []*Step{
		NewStep("load"),
		NewStep("transform"),
	}

	start := time.Now()

	for job := 0; job < jobCount; job++ {
		for _, step := range steps {
			step.Run()
		}
	}

	elapsed := time.Since(start).Microseconds()

	fmt.Printf("Running %d jobs through %d steps...\n", jobCount, len(steps))
	for _, step := range steps {
		fmt.Printf("Step %s processed %d jobs\n", step.name, step.processedCount)
	}
	fmt.Printf("Elapsed (microseconds): %d\n", elapsed)
}

File: languages/python/projects/04-expert/solutions/main.py
from __future__ import annotations

import time


class Step:
    def __init__(self, name: str) -> None:
        self._name = name
        self._processed_count = 0

    @property
    def name(self) -> str:
        return self._name

    @property
    def processed_count(self) -> int:
        return self._processed_count

    def run(self) -> None:
        self._processed_count += 1


def main() -> None:
    job_count = 3
    steps = [Step("load"), Step("transform")]

    start = time.perf_counter()

    for _ in range(job_count):
        for step in steps:
            step.run()

    elapsed_microseconds = int((time.perf_counter() - start) * 1_000_000)

    print(f"Running {job_count} jobs through {len(steps)} steps...")
    for step in steps:
        print(f"Step {step.name} processed {step.processed_count} jobs")
    print(f"Elapsed (microseconds): {elapsed_microseconds}")


if __name__ == "__main__":
    main()
